### PR TITLE
Doc comments: use std::unordered_map

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

As suggested in #11072 and now possible after #11092 

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/pos-idx.hh [hunk 1]
```diff
@@ -2,12 +2,15 @@
 
 #include <cinttypes>
 
+#include "util.hh"
```
Comment (fzakaria) [inline src/libexpr/pos-idx.hh:5]: Why is this included?
Can't seem to tell.

Comment (roberth) [reply src/libexpr/pos-idx.hh]: Thanks for asking. It was for `hash_combine`, but that wasn't obvious at all.
I've split that into its own header, partly because "util" is a bad scope for a header, but also for the less subjective argument, from the commit message:

> Splitting it out immediately answers questions like [this], without increasing the number of compilation units.
>
> I did consider using boost::hash_combine instead, but it doesn't seem to be quite as capable, accepting only two arguments.

File: src/libexpr/pos-idx.hh [hunk 2]
```diff
@@ -37,8 +40,28 @@ public:
     {
         return id == other.id;
     }
+
+    size_t hash() const noexcept
+    {
+        size_t h = 854125;
+        hash_combine(h, id);
```
Comment (edolstra) [inline src/libexpr/pos-idx.hh:47]: This `hash_combine()` seems unnecessary? We can just return the hash of `id`.

**Review by edolstra — APPROVED** [review]

File: src/libexpr/pos-idx.hh (hunk 2 above)
Comment (roberth) [reply src/libexpr/pos-idx.hh:47]: In Java it's common practice to include a random number to represent the distinct type, but I think you're right.
That doesn't really apply here, because we don't have a "top" type like Java's `Object` that would create a need to avoid collisions when used in e.g. `HashMap<String, Object>`.
The closest thing we might have is `std::variant`. I think we can assume that it hashes its index number, which is probably better than salting each type anyway, because iirc it allows multiple distinct occurrences of the same type.
